spring-boot-starter/starter-security/src/test/java/org/camunda/bpm/spring/boot/starter/security/oauth2/impl/CamundaIdentityProviderIT.java

```diff
@@ -0,0 +1,181 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.camunda.bpm.spring.boot.starter.security.oauth2.impl;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.atLeastOnce;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.mockConstruction;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.verify;
+import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.authentication;
+
+import org.camunda.bpm.engine.IdentityService;
+import org.camunda.bpm.engine.identity.Group;
+import org.camunda.bpm.engine.identity.GroupQuery;
+import org.camunda.bpm.engine.identity.User;
+import org.camunda.bpm.engine.identity.UserQuery;
+import org.camunda.bpm.engine.impl.identity.WritableIdentityProvider;
+import org.camunda.bpm.engine.impl.identity.db.DbGroupQueryImpl;
+import org.camunda.bpm.engine.impl.identity.db.DbUserQueryImpl;
+import org.camunda.bpm.spring.boot.starter.security.oauth2.AbstractSpringSecurityIT;
+import org.junit.Before;
+import org.junit.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.boot.test.web.client.TestRestTemplate;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.oauth2.client.OAuth2AuthorizedClientService;
+import org.springframework.security.oauth2.client.authentication.OAuth2AuthenticationToken;
+import org.springframework.security.oauth2.client.registration.ClientRegistrationRepository;
+import org.springframework.test.annotation.DirtiesContext;
+import org.springframework.test.context.TestPropertySource;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;
+import org.springframework.test.web.servlet.result.MockMvcResultHandlers;
+import org.springframework.test.web.servlet.result.MockMvcResultMatchers;
+
+@AutoConfigureMockMvc
+@TestPropertySource("/oauth2-mock.properties")
+@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)
+public class CamundaIdentityProviderIT extends AbstractSpringSecurityIT {
+
+  @Autowired
+  private IdentityService identityService;
+
+  @Autowired
+  private MockMvc mockMvc;
+
+  @Autowired
+  private TestRestTemplate restTemplate;
+
+  @Autowired
+  private ClientRegistrationRepository registrations;
+
+  @MockBean
```

Comment:
The `@MockBean` approach enforces us to use `@DirtiesContext` to re-create the application context from scratch each for each test because it replaces the original bean with a mock of the application context of the test.

I've tried to replace before / after each test the original bean with a Mock and restore the application context to its original state after each test. Unfortunately, i couldn't come up with a working example.

Reply:
Yes it's tricky to clean up the context manually, although it's definitely possible. On the other hand, testing each test independently from the rest with a non polluted context also seemed appealing.